services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -0,0 +1,415 @@
+        // apply to requests on this endpoint".
+        AuthorizeClientUtil.ClientAuthResult clientAuth = AuthorizeClientUtil.authorizeClient(session, event, null);
+        ClientModel client = clientAuth.getClient();
+
+        if (client == null) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST, Messages.MISSING_PARAMETER,
+                OIDCLoginProtocol.CLIENT_ID_PARAM);
+        }
+
+        checkClient(client.getClientId());
+
+        return client;
+    }
+
+    private ClientModel checkClient(String clientId) {
+        if (clientId == null) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST,
+                Messages.MISSING_PARAMETER, OIDCLoginProtocol.CLIENT_ID_PARAM);
+        }
+
+        event.client(clientId);
+
+        ClientModel client = realm.getClientByClientId(clientId);
+        if (client == null) {
+            event.error(Errors.CLIENT_NOT_FOUND);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST,
+                Messages.CLIENT_NOT_FOUND);
+        }
+
+        if (!client.isEnabled()) {
+            event.error(Errors.CLIENT_DISABLED);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST, Messages.CLIENT_DISABLED);
+        }
+
+        OIDCAdvancedConfigWrapper oidcAdvancedConfigWrapper = OIDCAdvancedConfigWrapper.fromClientModel(client);
+        if (!oidcAdvancedConfigWrapper.isOAuth2DeviceAuthorizationGrantEnabled()) {
+            event.error(Errors.NOT_ALLOWED);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST,
+                Messages.OAUTH2_DEVICE_AUTHORIZATION_GRANT_DISABLED);
+        }
+
+        if (client.isBearerOnly()) {
+            event.error(Errors.NOT_ALLOWED);
+            throw new ErrorPageException(session, null, Response.Status.FORBIDDEN, Messages.BEARER_ONLY);
+        }
+
+        String protocol = client.getProtocol();
+        if (protocol == null) {
+            logger.warnf("Client '%s' doesn't have protocol set. Fallback to openid-connect. Please fix client configuration",
+                clientId);
+            protocol = OIDCLoginProtocol.LOGIN_PROTOCOL;
+        }
+        if (!protocol.equals(OIDCLoginProtocol.LOGIN_PROTOCOL)) {
+            event.error(Errors.INVALID_CLIENT);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST, "Wrong client protocol.");
+        }
+
+        session.getContext().setClient(client);
+
+        return client;
+    }
+
+    public AuthenticationSessionModel createAuthenticationSession(ClientModel client) {
+        return createAuthenticationSession(client, (AuthorizationEndpointRequest) null);
+    }
+
+    public AuthenticationSessionModel createAuthenticationSession(ClientModel client, AuthorizationEndpointRequest request) {
```

Comment:
Some cleanup will be nice here. This method is never called with the "request" as an argument now. So we can probably just keep single method with the signature `createAuthenticationSession(ClientModel client)` and remove entirely the one with signature `createAuthenticationSession(ClientModel client, AuthorizationEndpointRequest request)` . Also we can remove all the blocks for `if (request != null)` .